File: 0110-balanced-binary-tree/0110-balanced-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
    let obj = {isBalanced: true};
    dfs(root, obj);
    return obj.isBalanced;
};

function dfs(root: TreeNode | null, obj: {isBalanced: boolean}): number {
    if (!root) return 0;
    const leftDepth = dfs(root.left, obj);
    const rightDepth = dfs(root.right, obj);
    if (Math.abs(leftDepth - rightDepth) > 1) obj.isBalanced = false;
    return Math.max(leftDepth + 1, rightDepth + 1);
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    balanced := true

    var depth func(tree *TreeNode) int
    depth = func(tree *TreeNode) int {
        if tree == nil {
            return 0
        }
        if tree.Left == nil && tree.Right == nil {
            return 1
        }

        leftDepth := depth(tree.Left)
        rightDepth := depth(tree.Right)

        determine := abs(rightDepth - leftDepth)
        if determine > 1 {
            balanced = false
        }
        if leftDepth > rightDepth {
            return leftDepth + 1
        }
        return rightDepth + 1
    }

    depth(root)
    return balanced
}

func abs(num1 int) int {
    if num1 < 0 {
        return -num1
    } else {
        return num1
    }
}